### PR TITLE
Web: Add autocompletion to repository compare page

### PR DESCRIPTION
**Pull request by umpox**
This PR:
- Adds autocomplete to repo compare page
- Updates the landing area when first loading the compare page
- Enables speculative search to support additional git revspec syntax
- Updates other copy and UI

https://user-images.githubusercontent.com/9516420/129737519-7988c8b5-adc2-4173-8c39-667b22263e0b.mov

### CODE REVIEW

**Review by eseliger — APPROVED** [review]
Love that change, it's so much better!

Two minor feedback items:

- For the pre-selected branch for base and head.. should we rather display a different body, ie. how to get started with git revspecs, or just as simple as select a rev above to compare? I find it a little weird that it is pre-populated and tries to load a diff, but it can technically never find one and it'll always just show that nothing found view when you load that page.
- Can we somehow advertise that the inputs support something that's NOT in the list of autocompleted items? Or rather, extend the autocomplete to better indicate it actually IS supported? Currently, it's not completely obvious to me that I can just type in something else. Three ideas:
  - Simply add a small notice somewhere saying git revspec is supported (and perhaps a docs link)
  - Add to the dropdown completion item something like a small help text such as `[main^] use revspec to filter commits`
  - Extend the dropdown completion thing a bit to understand some of these syntaxes and give a hint what the current expression does: `[main^] the parent commit of main`

**Comment by umpox** [comment]
Thanks @eseliger I agree, I've updated this to display a landing page with some short text and a link to the Git documentation. For now, I just link to the [Git docs](https://git-scm.com/docs/git-rev-parse.html#_specifying_revisions) don't believe we have any SG docs for this, unless I'm mistaken? I think it's worth adding our own docs for this in a later PR, it seems we don't support everything (e.g. `main@{yesterday}`) but it isn't immediately clear what we do/don't support here.

What do you think? I think this solution works for this initial approach. Adding additional context/understanding to certain commit nodes would be a great enhancement as the revspec docs aren't amazing. I also like how GitHub approaches this (e.g. at https://github.com/sourcegraph/sourcegraph/compare) by displaying a list of possibilities. We could easily do that too and could bring a lot of extra context to the user

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx [hunk 1]
```diff
@@ -1,124 +1,129 @@
...
-    private static BASE_INPUT_ID = 'repository-compare-header__base-spec'
-    private static HEAD_INPUT_ID = 'repository-compare-header__head-spec'
+export const RepositoryComparePopover: React.FunctionComponent<RepositoryComparePopoverProps> = ({
+    id,
+    comparison,
+    repo,
+    type,
+}) => {
+    const [popoverOpen, setPopoverOpen] = React.useState(false)
+    const togglePopover = React.useCallback(() => setPopoverOpen(previous => !previous), [])
```
Comment (valerybugakov) [inline client/web/src/repo/compare/RepositoryCompareHeader.tsx:47]: For consistency: 

```suggestion
    const togglePopover = useCallback(() => setPopoverOpen(previous => !previous), [])
```

Comment (valerybugakov) [reply client/web/src/repo/compare/RepositoryCompareHeader.tsx:47]: Is `useCallback` required here for perf reasons?

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx [hunk 2]
```diff
@@ -1,124 +1,129 @@
...
 
-    public render(): JSX.Element | null {
-        // Whether the user has entered new base/head values that differ from what's in the props and has not yet
-        // submitted the form.
-        const stateDiffers =
-            this.state.comparisonBaseSpec !== (this.props.base.revision || '') ||
-            this.state.comparisonHeadSpec !== (this.props.head.revision || '')
+            return `/${repo.name}/-/compare/${comparePath}`
+        },
+        [comparison, repo.name, type]
```
Comment (valerybugakov) [inline client/web/src/repo/compare/RepositoryCompareHeader.tsx:62]: It seems that the `comparison` prop is always a new object based on the markup [here](https://github.com/sourcegraph/sourcegraph/pull/23977/files#diff-a6440174439ccfc41a24f6cf55fd9b7f8c8aee10ca2c26c4ffe092c2566fc0e0R124): 

```tsx
<RepositoryComparePopover id="base-popover" type="base" comparison={{ base, head }} repo={repo} />
```

So `useCallback` can be omitted.

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx [hunk 3]
```diff
@@ -1,124 +1,129 @@
...
+    className,
+    repo,
+}) => (
+    <div className={`${styles.repositoryCompareHeader} ${className}`}>
+        <PageHeader
+            path={[{ text: 'Compare changes' }]}
+            description="Compare changes across revisions."
+            className="mb-3"
+        />
+        <div className={`${className}-inner d-flex align-items-center`}>
```
Comment (valerybugakov) [inline client/web/src/repo/compare/RepositoryCompareHeader.tsx:123]: The interpolated `className` results into `my-3-inner` currently. Should it be deleted?

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx [hunk 4]
```diff
@@ -1,124 +1,129 @@
...
+    )
 }
+
+export const RepositoryCompareHeader: React.FunctionComponent<RepositoryCompareHeaderProps> = ({
+    base,
+    head,
+    className,
+    repo,
+}) => (
+    <div className={`${styles.repositoryCompareHeader} ${className}`}>
```
Comment (valerybugakov) [inline client/web/src/repo/compare/RepositoryCompareHeader.tsx:117]: Can we use `classNames` util here?

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx [hunk 5]
```diff
@@ -1,124 +1,129 @@
...
 
-    private onChange: React.ChangeEventHandler<HTMLInputElement> = event => {
-        this.setState(
-            (event.currentTarget.id === RepositoryCompareHeader.BASE_INPUT_ID
-                ? { comparisonBaseSpec: event.currentTarget.value }
-                : { comparisonHeadSpec: event.currentTarget.value }) as Pick<
-                State,
-                'comparisonBaseSpec' & 'comparisonHeadSpec'
+    return (
+        <Button
```
Comment (valerybugakov) [inline client/web/src/repo/compare/RepositoryCompareHeader.tsx:74]: `Button` in the wild 🎉

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx [hunk 6]
```diff
@@ -1,124 +1,129 @@
...
+    repo: RepositoryCompareHeaderProps['repo']
 }
 
-/**
- * Header for the repository compare area.
- */
-export class RepositoryCompareHeader extends React.PureComponent<Props, State> {
-    private static BASE_INPUT_ID = 'repository-compare-header__base-spec'
-    private static HEAD_INPUT_ID = 'repository-compare-header__head-spec'
+export const RepositoryComparePopover: React.FunctionComponent<RepositoryComparePopoverProps> = ({
```
Comment (valerybugakov) [inline client/web/src/repo/compare/RepositoryCompareHeader.tsx:40]: Optional: Should this component live in a separate file?

**Review by valerybugakov — APPROVED** [review]
Thanks for adding short docs with a link to Git docs! Looks much better now.

**Comment by eseliger** [comment]
Looks good! Thanks for following up on that :)

**Comment by AlicjaSuska** [comment]
Thank you for sharing @umpox I've added some copy changes to this [Figma file](https://www.figma.com/file/TauSmn5u78AKN29wnlEund/Untitled?node-id=0%3A1)

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx (hunk 1 above)
Comment (umpox) [reply client/web/src/repo/compare/RepositoryCompareHeader.tsx:47]: Yeah, but more just habit really, putting this in `useCallback` won't have much of an overall effect. I'm trying to stop overusing this so going to change

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx (hunk 3 above)
Comment (umpox) [reply client/web/src/repo/compare/RepositoryCompareHeader.tsx:123]: Oh wow, good catch yeah this shouldn't be here

File: client/web/src/repo/compare/RepositoryCompareHeader.tsx (hunk 6 above)
Comment (umpox) [reply client/web/src/repo/compare/RepositoryCompareHeader.tsx:40]: Yes but I'm lazy 😁  I'll move it

**Pull request merged**